docs/tutorial/saxpy.rst

```diff
@@ -0,0 +1,780 @@
+:doc:`/install/install`.
+
+.. _hip-tutorial-saxpy-heterogeneous-programming
+
+Heterogeneous programming
+=========================
+
+*Heterogeneous programming* and *offloading APIs* are often mentioned together. Heterogeneous programming deals with devices of varying capabilities
+simultaneously. Offloading focuses on the "remote" and asynchronous aspects of
+computation. HIP encompasses both. It exposes GPGPU (general-purpose GPU)
+programming much like ordinary host-side CPU programming and lets you move data
+across various devices.
+
+When programming in HIP (and other heterogenous APIs for that matter), remember
+that target devices are built for a specific purpose. They are designed with
+different tradeoffs than traditional CPUs and therefore have very different
+performance characteristics. Even subtle changes in code might adversely affect
+execution time.
+
+Your first lines of HIP code
+============================
+
+First, let's do the "Hello, World!" of GPGPU: SAXPY. *SAXPY* is a mathematical
+acronym; a vector equation :math:`a\cdot x+y=z` where :math:`a\in\mathbb{R}` is
+a scalar and :math:`x,y,z\in\mathbb{V}` are vector quantities of some large
+dimensionality. This vector space is defined over the set of reals. Practically
+speaking, you can compute this using a single ``for`` loop over three arrays.
+
+.. code-block:: C++
+
+    for (int i = 0 ; i < N ; ++i)
+        z[i] = a * x[i] + y[i];
+
+In linear algebra libraries, such as BLAS (Basic Linear Algebra Subsystem) this
+operation is defined as AXPY "A times X Plus Y". The term SAXPY refers to the
+single-precision version of this operation
+
+The "S" comes from
+*single-precision*, meaning that array element is ``float`` -s
+(IEEE 754 binary32 representation).
+
+To quickly get started, use the set of `HIP samples from GitHub
+<https://github.com/amd/rocm-examples/>`_. With Git configured on your machine,
+open a command-line and navigate to your desired working directory, then run:
+
+.. code-block:: shell
+
+  git clone https://github.com/amd/rocm-examples.git
+
+A simple implementation of SAXPY resides in the ``HIP-Basic/saxpy/main.hip``
+file in this repository. The HIP code here mostly deals with where data
+has to be and when, and how devices transform this data. The first HIP calls
+deal with allocating device-side memory and dispatching data from host-side
+memory in a C runtime-like fashion.
+
+.. code-block:: C++
+
+  // Allocate and copy vectors to device memory.
+  float* d_x{};
+  float* d_y{};
+  HIP_CHECK(hipMalloc(&d_x, size_bytes));
+  HIP_CHECK(hipMalloc(&d_y, size_bytes));
+  HIP_CHECK(hipMemcpy(d_x, x.data(), size_bytes, hipMemcpyHostToDevice));
+  HIP_CHECK(hipMemcpy(d_y, y.data(), size_bytes, hipMemcpyHostToDevice));
+
+``HIP_CHECK`` is a custom macro borrowed from the examples utilities which
+checks the error code returned by API functions for errors and reports them to
+the console. It's not essential to the API.
```

Comment:
It is not essential to the API but it is a good practice to check the error codes of the HIP APIs incase you pass on incorrect values to the API or the API might be out of resources.

Reply:
Added.